src/util/unicode.cpp

```diff
@@ -224,43 +242,131 @@ Function: utf16_to_utf8
 
 \*******************************************************************/
 
-std::string utf16_to_utf8(const std::basic_string<unsigned short int> &s)
+const char **narrow_argv(int argc, const wchar_t **argv_wide)
 {
-  std::string result;
+  if(argv_wide==NULL)
+    return NULL;
 
-  result.reserve(s.size()); // at least that long
+  // the following never gets deleted
+  const char **argv_narrow=new const char *[argc+1];
+  argv_narrow[argc]=0;
 
-  for(const auto c : s)
-    utf32_to_utf8(c, result);
+  for(int i=0; i<argc; i++)
+    argv_narrow[i]=strdup(narrow(argv_wide[i]).c_str());
 
-  return result;
+  return argv_narrow;
 }
 
 /*******************************************************************\
 
-Function: narrow_argv
+Function: swap_last_bytes
 
-  Inputs:
+  Inputs: An integer
 
- Outputs:
+ Outputs: An integer whose representation swaps the last two bytes
 
- Purpose:
+ Purpose: A helper function for dealing with different UTF16  endians
 
 \*******************************************************************/
 
-const char **narrow_argv(int argc, const wchar_t **argv_wide)
+unsigned int swap_last_bytes(unsigned int x)
```

Comment:
Rather than write "byte-swap least-significant word in dword", might be clearer to write "byte-swap word" and have the caller cast to the smaller type.